### ISSUE TITLE
Architectural documentation

### ISSUE THREAD

**Issue by prettydiff** [prettydiff/prettydiff#38]
This project needs formal code/architecture/engineering documentation.

**Comment by Glavin001** [comment]
+1 Maybe then I could give you a hand with implementing some of these features / bug fixes :smiley:

**Comment by prettydiff** [comment]
Thank you so very much for the offer. This one is entirely on me though.  I need to spend time this next weekend diving into the code and describing the major parts and the various relationships and flow control of things.

**Comment by Glavin001** [comment]
Is there a way that we could make the code base for Node.js and the web browser library the same? We have been having a lot of these issues where features work on Prettydiff in the web app but not in the Node.js package.

**Comment by prettydiff** [comment]
Let me know of some of these features that are not working, and I will look into this. It is probably a code defect instead of not included features. It is likely there are bugs in the node file that I am not as quick to correct, because I primarily use prettydiff from a local file in the browser so that I can focus on interfacing and experimenting with the application file more quickly.

There is only one application which is contained the prettydiff.js file.

The tool in the web browser uses api/dom.js to run the webtool.  Most of the options and features are available from the webtool. The dom.js file has to account for storing settings into localStorage, the fileReader API, interacting with the DOM, and providing support for additional pages on the site.

The Node.js code is in api/node-local.js.  Every option is available in the api/node-local.js file. The Node version also contains additional functionality to read information from a variety of sources, which is determined by the *readmethod* option. The readmethod option is actually the majority of code in the api/node-local.js file because it provides support for recursively walking directory trees and in the case of the *diff* mode it has to correlate files in one directory tree against files in a different directory tree before perform a diff.

The only additional functionality provided to api/dom.js that is not provided to api/node-local.js is redundant language detection. The prettydiff.js already contains language detection so that such functionality is always present. When I added support for the *codemirror* application (the editor like line numbers and syntax coloring over my textareas) I had to add language detection to the dom.js file so that code is properly colored in the UI.

Lately I am considering removing codemirror. It throws a lot of errors and its code is too sloppy for any do-it-yourself repairs.  I have error reporting setup in the web tool (but only if accessed from my domain), which sends me an email about code errors.  About 70%+ of the errors I get are related to codemirror, 20% are related to users trying to do something hacky unrelated to my code, and the final 10% are my errors.
